### PR TITLE
session, statistics, util: fix data race of Handle.mu.ctx

### LINKED ISSUE #32867: DATA RACE in the StatementContext
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
==================
WARNING: DATA RACE
Write at 0x00c001ddb8f8 by goroutine 204:
  sync/atomic.CompareAndSwapInt32()
      /usr/local/go/src/runtime/race_amd64.s:334 +0xb
  sync/atomic.CompareAndSwapInt32()
      <autogenerated>:1 +0x1e
  github.com/pingcap/tidb/sessionctx/stmtctx.(*StatementContext).AppendWarnings()
      /go/tidb/sessionctx/stmtctx/stmtctx.go:563 +0x65
  github.com/pingcap/tidb/session.(*session).getInternalSession.func1()
      /go/tidb/session/session.go:1733 +0x449
  github.com/pingcap/tidb/session.(*session).withRestrictedSQLExecutor()
      /go/tidb/session/session.go:1765 +0x399
  github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL()
      /go/tidb/session/session.go:1769 +0x124
  github.com/pingcap/tidb/statistics/handle.(*Handle).execRestrictedSQL()
      /go/tidb/statistics/handle/handle.go:132 +0x188
  github.com/pingcap/tidb/statistics/handle.(*Handle).HandleUpdateStats()
      /go/tidb/statistics/handle/update.go:731 +0x84
  github.com/pingcap/tidb/domain.(*Domain).updateStatsWorker()
      /go/tidb/domain/domain.go:1451 +0x97a
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop·dwrap·19()
      /go/tidb/domain/domain.go:1304 +0x71
Previous write at 0x00c001ddb8f8 by goroutine 163:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).InitStatementContext()
      /go/tidb/sessionctx/variable/session.go:1019 +0x1f5
  github.com/pingcap/tidb/executor.ResetContextOfStmt()
      /go/tidb/executor/executor.go:1745 +0x13d
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1821 +0x396
  github.com/pingcap/tidb/session.(*session).ExecuteInternal()
      /go/tidb/session/session.go:1482 +0x484
  github.com/pingcap/tidb/statistics/handle.(*Handle).getStatsReader()
      /go/tidb/statistics/handle/handle.go:1449 +0x2f1
  github.com/pingcap/tidb/statistics/handle.(*Handle).getGlobalStatsReader()
      /go/tidb/statistics/handle/handle.go:1426 +0x125
  github.com/pingcap/tidb/statistics/handle.(*Handle).LoadNeededHistograms()
      /go/tidb/statistics/handle/handle.go:597 +0x6e
  github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker()
      /go/tidb/domain/domain.go:1369 +0x9ba
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop·dwrap·17()
      /go/tidb/domain/domain.go:1292 +0x39
Goroutine 204 (running) created at:
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1304 +0x470
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2874 +0x804
  github.com/pingcap/tidb/cmd/ddltest.createDDLSuite()
      /go/tidb/cmd/ddltest/ddl_test.go:128 +0x878
  github.com/pingcap/tidb/cmd/ddltest.TestColumn()
      /go/tidb/cmd/ddltest/column_test.go:118 +0x4a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
Goroutine 163 (running) created at:
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1292 +0x20a
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2874 +0x804
  github.com/pingcap/tidb/cmd/ddltest.createDDLSuite()
      /go/tidb/cmd/ddltest/ddl_test.go:128 +0x878
  github.com/pingcap/tidb/cmd/ddltest.TestColumn()
      /go/tidb/cmd/ddltest/column_test.go:118 +0x4a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
==================
WARNING: DATA RACE
Read at 0x00c001ddb948 by goroutine 204:
  github.com/pingcap/tidb/sessionctx/stmtctx.(*StatementContext).AppendWarnings()
      /go/tidb/sessionctx/stmtctx/stmtctx.go:565 +0xd0
  github.com/pingcap/tidb/session.(*session).getInternalSession.func1()
      /go/tidb/session/session.go:1733 +0x449
  github.com/pingcap/tidb/session.(*session).withRestrictedSQLExecutor()
      /go/tidb/session/session.go:1765 +0x399
  github.com/pingcap/tidb/session.(*session).ExecRestrictedSQL()
      /go/tidb/session/session.go:1769 +0x124
  github.com/pingcap/tidb/statistics/handle.(*Handle).execRestrictedSQL()
      /go/tidb/statistics/handle/handle.go:132 +0x188
  github.com/pingcap/tidb/statistics/handle.(*Handle).HandleUpdateStats()
      /go/tidb/statistics/handle/update.go:731 +0x84
  github.com/pingcap/tidb/domain.(*Domain).updateStatsWorker()
      /go/tidb/domain/domain.go:1451 +0x97a
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop·dwrap·19()
      /go/tidb/domain/domain.go:1304 +0x71
Previous write at 0x00c001ddb948 by goroutine 163:
  github.com/pingcap/tidb/sessionctx/variable.(*SessionVars).InitStatementContext()
      /go/tidb/sessionctx/variable/session.go:1019 +0x1f5
  github.com/pingcap/tidb/executor.ResetContextOfStmt()
      /go/tidb/executor/executor.go:1745 +0x13d
  github.com/pingcap/tidb/session.(*session).ExecuteStmt()
      /go/tidb/session/session.go:1821 +0x396
  github.com/pingcap/tidb/session.(*session).ExecuteInternal()
      /go/tidb/session/session.go:1482 +0x484
  github.com/pingcap/tidb/statistics/handle.(*Handle).getStatsReader()
      /go/tidb/statistics/handle/handle.go:1449 +0x2f1
  github.com/pingcap/tidb/statistics/handle.(*Handle).getGlobalStatsReader()
      /go/tidb/statistics/handle/handle.go:1426 +0x125
  github.com/pingcap/tidb/statistics/handle.(*Handle).LoadNeededHistograms()
      /go/tidb/statistics/handle/handle.go:597 +0x6e
  github.com/pingcap/tidb/domain.(*Domain).loadStatsWorker()
      /go/tidb/domain/domain.go:1369 +0x9ba
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop·dwrap·17()
      /go/tidb/domain/domain.go:1292 +0x39
Goroutine 204 (running) created at:
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1304 +0x470
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2874 +0x804
  github.com/pingcap/tidb/cmd/ddltest.createDDLSuite()
      /go/tidb/cmd/ddltest/ddl_test.go:128 +0x878
  github.com/pingcap/tidb/cmd/ddltest.TestColumn()
      /go/tidb/cmd/ddltest/column_test.go:118 +0x4a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
Goroutine 163 (running) created at:
  github.com/pingcap/tidb/domain.(*Domain).UpdateTableStatsLoop()
      /go/tidb/domain/domain.go:1292 +0x20a
  github.com/pingcap/tidb/session.BootstrapSession()
      /go/tidb/session/session.go:2874 +0x804
  github.com/pingcap/tidb/cmd/ddltest.createDDLSuite()
      /go/tidb/cmd/ddltest/ddl_test.go:128 +0x878
  github.com/pingcap/tidb/cmd/ddltest.TestColumn()
      /go/tidb/cmd/ddltest/column_test.go:118 +0x4a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
================== 
```

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

Issue Number: close #32867 #33001 #32822

Problem Summary:

We call `(*Handle).execRestrictedSQL` concurrently, which uses `Handle.mu.ctx` to fetch another session from session pool to execute internal SQL.
https://github.com/pingcap/tidb/blob/bd8c71018527c720afdc4acb480909639284323f/statistics/handle/handle.go#L132-L134
However, the method may cause data race in the following code.
https://github.com/pingcap/tidb/blob/bd8c71018527c720afdc4acb480909639284323f/session/session.go#L1748-L1764
At Line 1757, we append warnings for `Handle.mu.ctx`, which causes data race.

### What is changed and how it works?

IMO it is better to make `(*Handle).execRestrictedSQL`  inrelevant with `Handle.mu.ctx`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: statistics/handle/handle.go [hunk 1]
```diff
@@ -129,25 +128,43 @@ type Handle struct {
 	sysProcTracker sessionctx.SysProcTracker
 }
 
+func (h *Handle) withRestrictedSQLExecutor(ctx context.Context, fn func(context.Context, sqlexec.RestrictedSQLExecutor) ([]chunk.Row, []*ast.ResultField, error)) ([]chunk.Row, []*ast.ResultField, error) {
+	se, err := h.pool.Get()
+	if err != nil {
+		return nil, nil, errors.Trace(err)
+	}
+	defer h.pool.Put(se)
```
Comment (XuHuaiyu) [inline statistics/handle/handle.go:136]: Should we put this line before the `if` check?

**Review by hawkingrei — APPROVED** [review]

File: statistics/handle/handle.go (hunk 1 above)
Comment (chrysan) [reply statistics/handle/handle.go:136]: Should all internal sql executions using internal session pools ignore warnings?

Comment (xuyifangreeneyes) [reply statistics/handle/handle.go:136]: Currently we don't check warnings when executing internal sql through `(*Handle).execRestrictedSQL`, `(*Handle).execRestrictedSQLWithStatsVer` and `(*Handle).execRestrictedSQLWithSnapshot`. Maybe we can ignore warnings for now and make improvements if later warnings are needed.

File: statistics/handle/handle.go [hunk 2]
```diff
@@ -129,25 +128,43 @@ type Handle struct {
...
+
+	exec := se.(sqlexec.RestrictedSQLExecutor)
+	return fn(ctx, exec)
+}
+
 func (h *Handle) execRestrictedSQL(ctx context.Context, sql string, params ...interface{}) ([]chunk.Row, []*ast.ResultField, error) {
-	return h.mu.ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, []sqlexec.OptionFuncAlias{sqlexec.ExecOptionUseSessionPool}, sql, params...)
+	return h.withRestrictedSQLExecutor(ctx, func(ctx context.Context, exec sqlexec.RestrictedSQLExecutor) ([]chunk.Row, []*ast.ResultField, error) {
+		return exec.ExecRestrictedSQL(ctx, []sqlexec.OptionFuncAlias{sqlexec.ExecOptionUseCurSession}, sql, params...)
```
Comment (qw4990) [inline statistics/handle/handle.go:144]: It seems that we can decouple/detach `execRestrictedSQL` from `Handle` thoroughly in the future.

**Review by qw4990 — APPROVED** [review]
LGTM

**Comment by purelind** [comment]
### TiDB MergeCI notify
**🟠 Bad News! New failing [1] after this pr merged.**
These new failed integration tests seem to be caused by the current PR, please try to fix these new failed integration tests, thanks!
CI Name | Result | Duration | Compare with Parent commit
------------ | ------------- | ------------- | -------------
[idc-jenkins-ci/integration-cdc-test](https://ci.pingcap.net/job/tidb_ghpr_integration_cdc_test/1797/display/redirect) | 🟥 failed 12, success 22, total 34 | 27 min | New failing
[idc-jenkins-ci-tidb/integration-br-test](https://ci.pingcap.net/job/tidb_ghpr_integration_br_test/8764/display/redirect) | 🟥 failed 11, success 18, total 29 | 41 min | Existing failure
[idc-jenkins-ci-tidb/code-coverage](https://ci.pingcap.net/job/tidb_ghpr_coverage/2990/display/redirect) | ✅ Lines coverage: 63.48% | 13 min | Fixed
[idc-jenkins-ci-tidb/integration-copr-test](https://ci.pingcap.net/job/tidb_ghpr_integration_copr_test/9038/display/redirect) | ✅ all 1 tests passed | 12 min | Fixed
[idc-jenkins-ci-tidb/common-test](https://ci.pingcap.net/job/tidb_ghpr_common_test/9461/display/redirect) | ✅ all 12 tests passed | 10 min | Fixed
[idc-jenkins-ci-tidb/integration-compatibility-test](https://ci.pingcap.net/job/tidb_ghpr_integration_campatibility_test/10037/display/redirect) | ✅ all 1 tests passed | 10 min | Fixed
[idc-jenkins-ci-tidb/tics-test](https://ci.pingcap.net/job/tidb_ghpr_tics_test/8879/display/redirect) | 🟢 all 1 tests passed | 19 min | Existing passed
[idc-jenkins-ci-tidb/integration-common-test](https://ci.pingcap.net/job/tidb_ghpr_integration_common_test/10630/display/redirect) | 🟢 all 11 tests passed | 14 min | Existing passed
[idc-jenkins-ci-tidb/sqllogic-test-2](https://ci.pingcap.net/job/tidb_ghpr_sqllogic_test_2/9125/display/redirect) | 🟢 all 28 tests passed | 9 min 4 sec | Existing passed
[idc-jenkins-ci-tidb/sqllogic-test-1](https://ci.pingcap.net/job/tidb_ghpr_sqllogic_test_1/9187/display/redirect) | 🟢 all 26 tests passed | 8 min 30 sec | Existing passed
[idc-jenkins-ci-tidb/integration-ddl-test](https://ci.pingcap.net/job/tidb_ghpr_integration_ddl_test/9250/display/redirect) | 🟢 all 6 tests passed | 8 min 26 sec | Existing passed
[idc-jenkins-ci-tidb/mybatis-test](https://ci.pingcap.net/job/tidb_ghpr_mybatis/9049/display/redirect) | 🟢 all 1 tests passed | 3 min 37 sec | Existing passed
[idc-jenkins-ci-tidb/plugin-test](https://ci.pingcap.net/job/tidb_merged_pr_build/2859/display/redirect) | 🟢 build success, plugin test success | 4min | Existing passed

**Pull request merged**